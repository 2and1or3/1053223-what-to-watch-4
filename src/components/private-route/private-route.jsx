import React from "react";
import PropTypes from "prop-types";
import {Route, Redirect} from "react-router-dom";
import {connect} from "react-redux";

import {getUserStatus} from '../../reducer/user/selectors.js';
import {UserStatus, AppRoute} from '../../consts.js';

const PrivateRoute = (props) => {
  const {exact, path, render, userStatus} = props;

  const isAllow = userStatus === UserStatus.AUTH;

  return (
    <Route
      exact = {exact}
      path = {path}
      render = {() => {

        return (
          isAllow ? render() :
            <Redirect to = {AppRoute.LOGIN}/>
        );
      }}

    />
  );
};

PrivateRoute.propTypes = {
  exact: PropTypes.bool.isRequired,
  path: PropTypes.string.isRequired,
  render: PropTypes.func.isRequired,
  userStatus: PropTypes.string.isRequired,
};

const mapStateToProps = (state) => ({
  userStatus: getUserStatus(state),
});

const ConnectedPrivateRoute = connect(mapStateToProps, null)(PrivateRoute);

export {PrivateRoute};
export default ConnectedPrivateRoute;
