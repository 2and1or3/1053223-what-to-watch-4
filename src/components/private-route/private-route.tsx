import * as React from "react";
import {Route, Redirect, RouteProps} from "react-router-dom";
import {connect} from "react-redux";

import {getUserStatus} from '../../reducer/user/selectors';

type Props = RouteProps & {
  exact: boolean;
  path: string;
  render: (routeProps: RouteProps) => React.ReactNode;
  userStatus: string;
  allowForUserStatus: string;
  redirectTo: string;
}

const PrivateRoute: React.FunctionComponent<Props> = (props: Props) => {
  const {exact, path, render, userStatus, allowForUserStatus, redirectTo} = props;
  const isAllow = userStatus === allowForUserStatus;

  return (
    <Route
      exact = {exact}
      path = {path}
      render = {(routeProps) => {

        return (
          isAllow ? render(routeProps) :
            <Redirect to = {redirectTo}/>
        );
      }}

    />
  );
};

const mapStateToProps = (state) => ({
  userStatus: getUserStatus(state),
});

const ConnectedPrivateRoute = connect(mapStateToProps, null)(PrivateRoute);

export {PrivateRoute};
export default ConnectedPrivateRoute;
