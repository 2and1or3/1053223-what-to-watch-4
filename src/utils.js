import {AppRoute, MONTHS} from './consts';

const createVideoMock = (element) => {
  if (element.type === `video`) {
    return element;
  }

  return null;
};


const extend = (a, b) => {
  return Object.assign({}, a, b);
};

const adapterToLocalFilms = (films) => {

  const localFilms =
    films
    .slice()
    .map((film) => ({
      id: film.id.toString(),
      isFavorite: film.is_favorite,
      title: film.name,
      poster: film.preview_image,
      preview: film.preview_video_link,
      src: film.video_link,
      background: film.background_image,
      backgroundColor: film.background_color,
      cover: film.poster_image,
      genre: film.genre.toLowerCase(),
      release: film.released.toString(),
      description: film.description,
      rating: film.rating,
      voiceCount: film.scores_count,
      duration: film.run_time,
      director: film.director,
      actors: film.starring.slice(),
    }));

  return localFilms;
};

const adapterToLocalComments = (comments) => {

  const localComments =
  comments
  .map((comment) => ({
    id: comment.id.toString(),
    author: comment.user.name,
    date: comment.date,
    description: comment.comment,
    rate: comment.rating.toString(),
  }));

  return localComments;
};

const GetPath = {
  filmReview: (id) => `${AppRoute.FILM}${id}${AppRoute.REVIEW}`,
  filmPlayer: (id) => `${AppRoute.FILM}${id}${AppRoute.PLAYER}`,
  filmDetails: (id) => `${AppRoute.FILM}${id}`,
};

const getDateFormat = (date) => {
  const data = new Date(date);
  const month = data.getMonth();
  const day = data.getDate();
  const year = data.getFullYear();

  return `${MONTHS[month]} ${day} ${year}`;
};

const getWatchTimeFormat = (seconds) => {

  let hours = Math.floor(seconds / 3600);
  let min = Math.floor((seconds - hours * 3600) / 60);
  let sec = seconds - hours * 3600 - min * 60;

  hours = hours.toString();
  min = min.toString().padStart(2, `0`);
  sec = sec.toString().padStart(2, `0`);

  return `${hours}:${min}:${sec}`;
};

const capitalize = (str) => {
  str = str.toLowerCase();
  let words = str.split(` `);

  words = words.map((word) => word[0].toUpperCase() + word.slice(1));
  str = words.join(` `);

  return str;
};

const getUniqueGenres = (genres) => {
  let uniqueGenres = [];

  genres.forEach((genre) => {
    const isRepeated = uniqueGenres.some((uniqueGenre) => uniqueGenre === genre);

    if (!isRepeated) {
      uniqueGenres.push(genre);
    }
  });


  const localUniqueGenres = uniqueGenres.map((uniqueGenre) => ({
    id: uniqueGenre,
    title: capitalize(uniqueGenre),
  }));

  return localUniqueGenres;
};

const noop = () => {};

export {createVideoMock, adapterToLocalFilms, extend, adapterToLocalComments, GetPath, getDateFormat, getWatchTimeFormat, getUniqueGenres, noop};
