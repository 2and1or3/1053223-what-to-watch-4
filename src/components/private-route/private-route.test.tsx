import * as React from "react";
import * as renderer from "react-test-renderer";
import {BrowserRouter} from "react-router-dom";

import {PrivateRoute} from './private-route';
import {noop} from '../../utils';

it(`Render PrivateRoute component`, () => {
  const tree = renderer
    .create(
        <BrowserRouter>
          <PrivateRoute
            exact
            path = {`some path`}
            render = {noop}
            userStatus = {`AUTH`}
            allowForUserStatus = {`AUTH`}
            redirectTo = {`/login`}
          />
        </BrowserRouter>
    ).toJSON();

  expect(tree).toMatchSnapshot();
});

// npm run test.jest -- components/private-route/private-route.test
