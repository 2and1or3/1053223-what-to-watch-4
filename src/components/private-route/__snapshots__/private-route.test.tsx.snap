// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render PrivateRoute component 1`] = `null`;
